test/bio_base64_test.c

```diff
@@ -6,107 +6,263 @@
+
+        return BIO_write(mem, encoded, elen) == elen;
+    }
+
+    /* Encode full 3-octet groups */
+    while (buflen > 2) {
+        unsigned long v = buf[0] << 16 | buf[1] << 8 | buf[2];
+
+        if (memoutws(mem, b64[v >> 18], wscnt, llen, &pos) == 0
+            || memoutws(mem, b64[(v >> 12) & 0x3f], wscnt, llen, &pos) == 0
+            || memoutws(mem, b64[(v >> 6) & 0x3f], wscnt, llen, &pos) == 0
+            || memoutws(mem, b64[v & 0x3f], wscnt, llen, &pos) == 0)
+            return 0;
+        buf += 3;
+        buflen -= 3;
+    }
+
+    /* Encode and pad final 1 or 2 octet group */
+    if (buflen == 2) {
+        unsigned long v = buf[0] << 8 | buf[1];
+
+        if (memoutws(mem, b64[(v >> 10) & 0x3f], wscnt, llen, &pos) == 0
+            || memoutws(mem, b64[(v >> 4) & 0x3f], wscnt, llen, &pos) == 0
+            || memoutws(mem, b64[(v & 0xf) << 2], wscnt, llen, &pos) == 0
+            || memoutws(mem, '=', wscnt, llen, &pos) == 0)
+            return 0;
+    } else if (buflen == 1) {
+        unsigned long v = buf[0];
+
+        if (memoutws(mem, b64[v >> 2], wscnt, llen, &pos) == 0
+            || memoutws(mem, b64[(v & 0x3) << 4], wscnt, llen, &pos) == 0
+            || memoutws(mem, '=', wscnt, llen, &pos) == 0
+            || memoutws(mem, '=', wscnt, llen, &pos) == 0)
+            return 0;
+    }
+
+    /* Terminate last line */
+    if (pos > 0 && BIO_write(mem, &nl, 1) != 1)
+        return 0;
+
+    return 1;
+}
+
+static int genb64(char *prefix, char *suffix, unsigned const char *buf,
+                  unsigned buflen, char *encoded, unsigned llen,
+                  unsigned wscnt, char **out)
+{
+    int preflen = strlen(prefix);
+    int sufflen = strlen(suffix);
+    int outlen;
+    BUF_MEM *bptr;
+    BIO *mem = BIO_new(BIO_s_mem());
+
+    if (mem == NULL)
+        return -1;
+
+    if (BIO_write(mem, prefix, preflen) != preflen
+        || encode(buf, buflen, encoded, llen, wscnt, mem) <= 0
+        || BIO_write(mem, suffix, sufflen) != sufflen) {
+        BIO_free(mem);
+        return -1;
+    }
+
+    /* Orphan the memory BIO's data buffer */
+    BIO_get_mem_ptr(mem, &bptr);
+    *out = bptr->data;
+    outlen = bptr->length;
+    bptr->data = NULL;
+    (void) BIO_set_close(mem, BIO_NOCLOSE);
+    BIO_free(mem);
+    BUF_MEM_free(bptr);
+
+    return outlen;
+}
+
+static int single_test(test_case *t, int eof_return, int llen, int wscnt)
 {
-    char *out;
+    unsigned char *raw;
+    unsigned char *out;
+    unsigned out_len;
+    char *encoded = NULL;
+    int elen;
     BIO *bio, *b64;
-    int out_len;
     int n, n1, n2;
     int ret;
 
-    bio = BIO_new(BIO_s_mem());
+    /*
+     * Pre-encoded data always encodes NUL octets.  If all we care about is the
+     * length, and not the payload, use random bytes.
+     */
+    if (t->encoded != NULL)
+        raw = OPENSSL_zalloc(t->bytes);
+    else
+        raw = genbytes(t->bytes);
+
+    if (raw == NULL && t->bytes > 0)
```

Comment:
this may not be relevant, but do we have any platforms that return a non-null pointer when an allocation of size 0 is requested?

Reply:
I don't know whether any platforms instead return a fixed non-null pointer for malloc(0), it is certainly compliant with standards. The code will work either way.